swap/int256/int256.go

```diff
@@ -0,0 +1,161 @@
+// any int64 is valid as a Int256
+func Int256From(base int64) *Int256 {
+	u := new(Int256)
+	u.value = new(big.Int).SetInt64(base)
+	return u
+}
+
+// Copy creates and returns a new Int256 instance, with its underlying value set matching the receiver
+func (u *Int256) Copy() *Int256 {
+	v := new(Int256)
+	v.value = new(big.Int).Set(u.value)
+	return v
+}
+
+// Value returns the underlying private value for a Int256 struct
+func (u *Int256) Value() *big.Int {
+	return new(big.Int).Set(u.value)
+}
+
+// set assigns the underlying value of the given Int256 param to u, and returns the modified receiver struct
+// returns an error when the value cannot be correctly set
+func (u *Int256) set(value *big.Int) (*Int256, error) {
+	if err := checkInt256Bounds(value); err != nil {
+		return nil, err
+	}
+	if u.value == nil {
+		u.value = new(big.Int)
+	}
+	u.value.Set(value)
+	return u, nil
+}
+
+// checkInt256Bounds returns an error when the given value falls outside of the signed 256-bit integer range
+// returns nil otherwise
+func checkInt256Bounds(value *big.Int) error {
+	if value.Cmp(maxInt256) == 1 {
+		return fmt.Errorf("cannot set Int256 to %v as it overflows max value of %v", value, maxInt256)
+	}
+	if value.Cmp(minInt256) == -1 {
+		return fmt.Errorf("cannot set Int256 to %v as it underflows min value of %v", value, minInt256)
+	}
+	return nil
+}
+
+// Add sets u to augend + addend and returns u as the sum
+// returns an error when the value cannot be correctly set
+func (u *Int256) Add(augend, addend *Int256) (*Int256, error) {
+	sum := new(big.Int).Add(augend.value, addend.value)
+	return u.set(sum)
+}
+
+// Sub sets u to minuend - subtrahend and returns u as the difference
+// returns an error when the value cannot be correctly set
+func (u *Int256) Sub(minuend, subtrahend *Int256) (*Int256, error) {
+	difference := new(big.Int).Sub(minuend.value, subtrahend.value)
+	return u.set(difference)
+}
+
+// Mul sets u to multiplicand * multiplier and returns u as the product
+// returns an error when the value cannot be correctly set
+func (u *Int256) Mul(multiplicand, multiplier *Int256) (*Int256, error) {
+	product := new(big.Int).Mul(multiplicand.value, multiplier.value)
+	return u.set(product)
+}
+
+// cmp calls the underlying Cmp method for the big.Int stored in a Int256 struct as its value field
+func (u *Int256) cmp(v *Int256) int {
+	return u.value.Cmp(v.value)
+}
+
+// Equals returns true if the two Int256 structs have the same underlying values, false otherwise
+func (u *Int256) Equals(v *Int256) bool {
+	return u.cmp(v) == 0
+}
+
+// String returns the string representation for Int256 structs
+func (u *Int256) String() string {
+	return u.value.String()
+}
+
+// MarshalJSON implements the json.Marshaler interface
+// it specifies how to marshal a Int256 struct so that it can be written to disk
+func (u *Int256) MarshalJSON() ([]byte, error) {
+	return []byte(u.value.String()), nil
```

Comment:
This returns a json number (no quotes for json string). Being a number is it ok to overflow 64 bits resulting in potential interoperability problems. A few notes on that is in RFC 7159 section 6 Numbers.

This may be completely fine but would require a warning on decoding if someone uses Int256 externally, like in json http responses. Given that this package and json marshalling is available for any usage.

Additionally, big.Int.String can return `<nil>` string value, which should be replaced with `null` to comply with Int256.UnmarshalJSON.

Reply:
**regarding the json number issue:**

i've added quoting and unquoting calls to the marshalling and unmarshalling functions of both the `Int256` and `Uint256` structs.

**regarding the nil issue:** 

your comment made me realize we weren't checking in a strict manner whether these structs were allowed to have their values set to `nil`. i've changed the checking functions to return an error in this case, and added test cases for it.

as a result, i've removed the block of code which checks for `string(b) == "null"` in the `UnmarshalJSON` func. both this and the `DecodeRLP` funcs have calls which will output errors when attempting to set any `Int256` or `Uint256` `value` field to `nil`.